File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode merge = null;

        ListNode current1 = list1;
        ListNode current2 = list2;

        while (current1 != null && current2 != null) {
            int val1 = current1.val;
            int val2 = current2.val;

            if (val1 <= val2) {
                merge = insert(merge, val1);
                current1 = current1.next; 
            } else {
                merge = insert(merge, val2);
                current2 = current2.next; 
            }
        }


        while (current1 != null) {
            merge = insert(merge, current1.val);
            current1 = current1.next;
        }

        while (current2 != null) {
            merge = insert(merge, current2.val);
            current2 = current2.next;
        }

        return merge;
    }

    public ListNode insert(ListNode list, int data) {
        ListNode newNode = new ListNode(data);

        if (list == null) {
            return newNode;
        }

        ListNode current = list;
        while (current.next != null) {
            current = current.next; 
        }

        current.next = newNode; 
        return list;
    }
}

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        if list1 is None:
            return list2
        elif list2 is None:
            return list1
        
        if list1.val > list2.val: #list2 is smaller
            head = list2
            list2 = list2.next
        else:
            head = list1
            list1 = list1.next
        curr = head
        while list1 != None and list2 != None:
            if list1.val > list2.val: #list2 is smaller
                curr.next = list2
                list2 = list2.next
                curr = curr.next
            else:
                curr.next = list1
                list1 = list1.next
                curr = curr.next
        if list1 is not None:
            curr.next = list1
        else:
            curr.next = list2
        return head


        